File: bms/go/geo/geo_s2_CellID_distanceFromBegin/geo_s2_CellID_distanceFromBegin.go
package test

import "math/bits"

const (
	geo_s2_MaxLevel = 30
)

func (ci geo_s2_CellID) distanceFromBegin() int64 {
	return int64(ci >> uint64(2*(geo_s2_MaxLevel-ci.Level())+1))
}

func (ci geo_s2_CellID) Level() int {
	return geo_s2_MaxLevel - geo_s2_findLSBSetNonZero64(uint64(ci))>>1
}

func geo_s2_findLSBSetNonZero64(x uint64) int {
	if x == 0 {
		return 0
	}
	return bits.TrailingZeros64(x)
}

type geo_s2_CellID uint64

File: translations_only/fail/go/geo/geo_s2_CellID_distanceFromBegin/gpt4/Restart/geo_s2_CellID_distanceFromBegin.rs

mod test {
    const GEO_S2_MAX_LEVEL: i32 = 30;

    pub struct GeoS2CellID(u64);

    impl GeoS2CellID {
        pub fn distance_from_begin(&self) -> i64 {
            (self.0 >> (2 * (GEO_S2_MAX_LEVEL - self.level()) + 1)) as i64
        }

        pub fn level(&self) -> i32 {
            GEO_S2_MAX_LEVEL - geo_s2_find_lsb_set_non_zero64(self.0) >> 1
        }
    }

    pub fn geo_s2_find_lsb_set_non_zero64(x: u64) -> i32 {
        if x == 0 {
            0
        } else {
            x.trailing_zeros() as i32
        }
    }
}
